### ISSUE TITLE
Compiler does not recognize an uninitialized instance variable when a constructor from a superclass is involved

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#4151]
Considering the following two classes (crystal 0.21.0):
```crystal
class Foo
  property other : String
  property size : Int32
  def initialize(other)
    initialize(other, other.size)
  end
  
  def initialize(@other, @size)
  end
end

class Bar < Foo
  property value : String
  
  def initialize(value)
    initialize("hi", 2)
  end
  
  def initialize(@value, other : String)
    super(other)
  end
end

Bar.new("break!")
```

The compiler should complain that the instance variable `@value` is not set in the first constructor of `Bar`, but it doesn't. Instead the code compiles and results in an invalid memory access at runtime.

I guess the compiler wrongly assumes the call to `initialize("hi", 2)` would invoke the second constructor of `Bar` and thus initialize the instance variable. It seems it does not realize the type restrictions result in the constructor of `Foo` being invoked instead and `@value` will not be initialized.

This might be related to #4147 which I filed earlier today. But I encountered both issues in different circumstances.

**Comment by kirbyfan64** [comment]
Here's a slightly shorter test case:

```crystal
class Foo
  def initialize(x : String, y : Int32)
  end
end

class Bar < Foo
  getter value
  
  def initialize
    initialize("x", 1)
  end
  
  def initialize(@value : String, y : String)
  end
end

puts Bar.new.value
```

**Comment by vlazar** [comment]
In Crystal 0.31.1 on MacOS

```
$ crystal --version
Crystal 0.31.1 (2019-10-02)

LLVM: 8.0.1
Default target: x86_64-apple-macosx
```

[Johannes' original example](https://github.com/crystal-lang/crystal/issues/4151#issue-214515676) is not producing invalid memory access at runtime anymore. The output is empty.

[Ryan's example](https://github.com/crystal-lang/crystal/issues/4151#issuecomment-286882750) however still causes Invalid memory access.

**Comment by straight-shoota** [comment]
Yeah it's the `.value` call. Adding that in my example produces invalid memory access as well.

**Comment by vlazar** [comment]
Can you update your example please to avoid confusion?
